src/main/java/org/javarosa/core/model/TriggerableDag.java

```diff
@@ -365,45 +365,37 @@ public void finalizeTriggerables(FormInstance mainInstance, EvaluationContext ev
         return edges;
     }
 
-    public static Set<QuickTriggerable> buildDag(Set<QuickTriggerable> unorderedTriggerables, List<QuickTriggerable[]> edges) {
+    private static Set<QuickTriggerable> buildDag(Set<QuickTriggerable> vertices, List<QuickTriggerable[]> edges) {
         Set<QuickTriggerable> dag = new LinkedHashSet<>();
         // TODO Study how this algorithm ensures that we follow the ancestor >>> descendant direction and if the sorting step is strictly required
 
-        // DAGify the triggerables based on dependencies and sort them so that
-        // triggerables come only after the triggerables they depend on
-        List<QuickTriggerable> vertices = new ArrayList<>(unorderedTriggerables);
+        Set<QuickTriggerable> remainingVertices = new HashSet<>(vertices);
+        Set<QuickTriggerable[]> remainingEdges = new HashSet<>(edges);
+        while (remainingVertices.size() > 0) {
 
-        List<QuickTriggerable> orderedRoots = new ArrayList<>();
-        Set<QuickTriggerable> roots = new HashSet<>(vertices.size());
-        while (vertices.size() > 0) {
-            // determine root nodes
-            roots.clear();
-            roots.addAll(vertices);
-            for (QuickTriggerable[] edge : edges) {
+            Set<QuickTriggerable> roots = new HashSet<>(remainingVertices);
+            for (QuickTriggerable[] edge : remainingEdges)
                 roots.remove(edge[1]);
-            }
 
-            // if no root nodes while graph still has nodes, graph has cycles
             if (roots.size() == 0)
                 throwCyclesInDagException(vertices);
 
-            // order the root nodes - so the order is fixed
-            orderedRoots.clear();
-            orderedRoots.addAll(roots);
+            List<QuickTriggerable> orderedRoots = new ArrayList<>(roots);
             // TODO Study if insertion order & using LinkedSets would suffice to ensure proper triggerable chaining
             Collections.sort(orderedRoots, QuickTriggerableComparator.INSTANCE);
-
-            // remove root nodes and edges originating from them
-            // add them to the triggerablesDAG.
-            for (QuickTriggerable root : orderedRoots) {
-                dag.add(root);
-                vertices.remove(root);
-            }
-            for (int i = edges.size() - 1; i >= 0; i--) {
-                QuickTriggerable[] edge = edges.get(i);
-                if (roots.contains(edge[0]))
-                    edges.remove(i);
-            }
+            dag.addAll(orderedRoots);
+
+            Set<QuickTriggerable> newRemainingVertices = new HashSet<>();
+            for (QuickTriggerable vertex : vertices)
+                if (!dag.contains(vertex))
```

Comment:
This feels hard to follow to me. How about `remainingVertices.removeAll(orderedRoots)`? I know the goal was to remove mutation but in this case the mutating approach seems easier to understand. `remainingVertices.removeAll(orderedRoots)` feels nicely symmetric with `dag.addAll(orderedRoots)` so it feels like the roots are moving from one collection to the other.

Reply:
> it feels like the roots are moving from one collection to the other.

This is a very nice way of thinking about what's happening here. It didn't occur to me that it could be explained that way, and I think it would make the algorithm easier to understand which makes it worth the mutation of that collection.